pkg/image/admission/imagepolicy/imagepolicy.go

```diff
@@ -0,0 +1,392 @@
+
+	resolveGroupResources []unversioned.GroupResource
+
+	resolver imageResolver
+}
+
+// TODO: add wants the image registry resolver
+var _ = oadmission.WantsOpenshiftClient(&imagePolicyPlugin{})
+var _ = oadmission.Validator(&imagePolicyPlugin{})
+
+type integratedRegistryMatcher struct {
+	rules.RegistryMatcher
+}
+
+type imageResolver interface {
+	Resolve(ref imageapi.DockerImageReference) (*imageapi.Image, error)
+}
+
+// imagePolicyPlugin returns an admission controller for pods that controls what images are allowed to run on the
+// cluster.
+func newImagePolicyPlugin(client clientset.Interface, parsed *api.ImagePolicyConfig) (*imagePolicyPlugin, error) {
+	m := integratedRegistryMatcher{
+		RegistryMatcher: rules.NewRegistryMatcher(nil),
+	}
+	accepter := rules.NewExecutionRulesAccepter(parsed.ExecutionRules, m)
+	resourceAdjuster := rules.NewConsumptionRulesAdjuster(parsed.ConsumptionRules, m)
+	placementAdjuster := rules.NewPlacementRulesAdjuster(parsed.PlacementRules, m)
+
+	return &imagePolicyPlugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		config:  parsed,
+
+		accepter: accepter,
+		adjuster: rules.Adjusters{resourceAdjuster, placementAdjuster},
+
+		integratedRegistryMatcher: m,
+	}, nil
+}
+
+func (a *imagePolicyPlugin) SetDefaultRegistryFunc(fn imageapi.DefaultRegistryFunc) {
+	a.integratedRegistryMatcher.RegistryMatcher = rules.RegistryNameMatcher(fn)
+}
+
+func (a *imagePolicyPlugin) SetOpenshiftClient(c client.Interface) {
+	a.client = c
+}
+
+// Validate ensures that all required interfaces have been provided, or returns an error.
+func (a *imagePolicyPlugin) Validate() error {
+	if a.client == nil {
+		return fmt.Errorf("%s needs an Openshift client", api.PluginName)
+	}
+	imageResolver, err := newImageResolutionCache(a.client.Images(), a.client, a.integratedRegistryMatcher)
+	if err != nil {
+		return fmt.Errorf("unable to create image policy controller: %v", err)
+	}
+	a.resolver = imageResolver
+	return nil
+}
+
+// Admit attempts to apply the image policy to the incoming resource.
+func (a *imagePolicyPlugin) Admit(attr admission.Attributes) error {
+	switch attr.GetOperation() {
+	case admission.Create, admission.Update:
+		if len(attr.GetSubresource()) > 0 {
+			return nil
+		}
+		// only create and update are tested, and only on core resources
+		// TODO: scan all resources
+		// TODO: Create a general equivalence map for admission - operation X on subresource Y is equivalent to reduced operation
+	default:
+		return nil
+	}
+
+	gr := attr.GetResource().GroupResource()
+	accept, adjust := a.accepter.Covers(gr), a.adjuster.Covers(gr)
+	if !accept && !adjust {
+		return nil
+	}
+
+	podSpec, basePath, err := meta.GetPodSpec(attr.GetObject())
+	if err != nil {
+		return apierrs.NewForbidden(gr, attr.GetName(), fmt.Errorf("unable to apply image policy against objects of type %T, because they do not have a pod spec", attr.GetObject()))
+	}
+
+	covered := make(policyAttributes)
```

Comment:
> What about resolving and reassigning the container.Image here?

that would make it really easy to split off into a separate unit if someone wanted.  I could see it having some general utility.